### PR TITLE
feat(sequencer-relayer)!: provide filter for rollups

### PR DESCRIPTION
**Pull request by Fraser999**
## Summary
A new configuration env var has been added to allow filtering data submission to a subset of rollups.

## Background
This allows more flexibility in what is actually submitted to the Celestia chain.

## Changes
A new env var was added, mapped to a new field `Config::rollup_id_filter`.  This represents a list of rollups whose transactions should be included in blobs submitted to Celestia.  Rollups not in the list do not have their data submitted.  However, if the list is empty, no filtering is applied; all rollups have their data submitted.

The collection of filtered rollup IDs is passed down to the `write::conversion::convert` function where sequencer blocks are converted to blobs.  The blobs are filtered inside this function, partly since this is relatively early in the flow, but also so that the `ConversionInfo` struct can carry information about blobs being excluded.  This data is logged at the start of `write::submit_blobs` in an info-level log message.  It might be worthwhile adding metrics around the number of excluded rollups per submission, but I didn't see a clear benefit to that and it can be added in a follow-up PR if required.

## Testing
- Added unit tests covering parsing of the config string to a collection of `RollupId`s.
- ~Added a black box test where a sequencer block is constructed with several transactions, half from included rollups and half from excluded ones.  Existing black box tests ensure that when the filter is empty, no filtering is done.~ These all got removed in #963, and should be restored as part of #1008.

## Breaking Changelist
- Added `ASTRIA_SEQUENCER_RELAYER_ONLY_INCLUDE_ROLLUPS` which can be empty or can be a comma-separated list.

### CODE REVIEW

File: crates/astria-sequencer-relayer/tests/blackbox/helper.rs
Comment (noot) [inline crates/astria-sequencer-relayer/tests/blackbox/helper.rs:316]: just flagging that i updated these tests to use `SequencerBlock`s into of cometbft blocks, will need to be updated

**Review by noot — APPROVED** [review]
looks good!

File: crates/astria-sequencer-relayer/src/config.rs
Comment (SuperFluffy) [inline crates/astria-sequencer-relayer/src/config.rs:152]: My preference for the sort of tests is to `unwrap_err` instead of asserting on a boolean. If this fails then the person for whom this failed (i.e. `create_filter` returned with success) will have to start rewriting the test to figure out why this is `false`

File: crates/astria-sequencer-relayer/src/config.rs
Comment (SuperFluffy) [inline crates/astria-sequencer-relayer/src/config.rs:156]: Same as above, prefer `unwrap_err` over asserting on booleans for UX reasons.

File: crates/astria-sequencer-relayer/src/relayer/write/conversion.rs
Comment (SuperFluffy) [inline crates/astria-sequencer-relayer/src/relayer/write/conversion.rs:66]: I understand the design decision here, but I am worried that this might quickly become confusing a few months down the line.

Instead of passing this `HashSet<RollupId>` around, could this be made into a newtype wrapper `IncludeRollup(HashSet<RollupId>)` with a single `IncludeRollup::include: rollup_id -> bool` that contains this logic?

Extra idea: if we make this `IncludeRollup<Arc<HashSet<RollupId>>` then this can be cheaply cloned (and the `needless_pass_by_value` avoided?).

File: crates/astria-sequencer-relayer/src/relayer/mod.rs
Comment (SuperFluffy) [inline crates/astria-sequencer-relayer/src/relayer/mod.rs:88]: Could this be made a `Arc<HashSet<RollupId>>` since this will never be mutated again? This way we can avoid the (moderatly expensive) clones.

File: crates/astria-sequencer-relayer/local.env.example
Comment (SuperFluffy) [inline crates/astria-sequencer-relayer/local.env.example:47]: We should probably explicitly spell out that rollup IDs are 32 byte hashes, in case somebody is confused (maybe "often the sha256 hash of a rollup name").

**Review by SuperFluffy — COMMENTED** [review]
I agree with the direction of this PR.

The only thing I am worried about is the relatively complex use of `filter.is_empty() || filter.contains(&rollup_id)` to test if a block should be permitted. Could we make this into a dedicated type with a single `filter -> bool` method instead?

This feels like the classic thing that is unclear a few months down (plus this gives us the chance to make all these clones cheaper).

File: crates/astria-sequencer-relayer/tests/blackbox/helper.rs
Comment (Fraser999) [reply crates/astria-sequencer-relayer/tests/blackbox/helper.rs:316]: Thanks!  Unfortunately a recent PR deleted most of these tests, but #1008 will see them restored.  I'll merge our changes so when the tests are restored, we can reintroduce the test I initially added in this PR, but which has been deleted after merging `main`.

File: crates/astria-sequencer-relayer/src/relayer/write/conversion.rs
Comment (Fraser999) [reply crates/astria-sequencer-relayer/src/relayer/write/conversion.rs:66]: The newtype makes sense to me, but the `Arc` seems a little overkill given I'm expecting the set to be empty or at most have one or two entries, with the entries being `Copy` types.  It's also only being cloned once per sequencer block.  Saying that, I guess the `Arc` can't hurt, so I'll add it :)

The `needless_pass_by_value` still won't be avoided though, since Clippy wants me to change the signature to take a ref to the set, and the lifetime on it would need to be `'static` if I did that.

File: crates/astria-sequencer-relayer/tests/blackbox/helper.rs
Comment (Fraser999) [reply crates/astria-sequencer-relayer/tests/blackbox/helper.rs]: Done in [81686cc](https://github.com/astriaorg/astria/pull/1001/commits/81686cc4a3d4e1540b2f594e06e622b6dc597dc8).

File: crates/astria-sequencer-relayer/src/config.rs
Comment (Fraser999) [reply crates/astria-sequencer-relayer/src/config.rs]: Done in [aa8957e](https://github.com/astriaorg/astria/pull/1001/commits/aa8957e8c1a7a2d49742e755f3c4b0617fbda1e9).

File: crates/astria-sequencer-relayer/src/config.rs
Comment (Fraser999) [reply crates/astria-sequencer-relayer/src/config.rs]: Done in [aa8957e](https://github.com/astriaorg/astria/pull/1001/commits/aa8957e8c1a7a2d49742e755f3c4b0617fbda1e9).

File: crates/astria-sequencer-relayer/src/relayer/write/conversion.rs
Comment (Fraser999) [reply crates/astria-sequencer-relayer/src/relayer/write/conversion.rs]: Done in [02b1276](https://github.com/astriaorg/astria/pull/1001/commits/02b1276267c6ac65b72c78191ebecdfae92e72d9).

File: crates/astria-sequencer-relayer/src/relayer/mod.rs
Comment (Fraser999) [reply crates/astria-sequencer-relayer/src/relayer/mod.rs]: Done in [02b1276](https://github.com/astriaorg/astria/pull/1001/commits/02b1276267c6ac65b72c78191ebecdfae92e72d9).

File: crates/astria-sequencer-relayer/local.env.example
Comment (Fraser999) [reply crates/astria-sequencer-relayer/local.env.example]: Done in [cb1238b](https://github.com/astriaorg/astria/pull/1001/commits/cb1238b701b520d1bdbd778f544777afb9242736).

**Review by SuperFluffy — APPROVED** [review]
Looks great, thank you for the PR.

I am not sure about the state of the blackbox since #963 removed a ton of that functionality.

Since #1008 is a follow-up PR that is imminent I would ask for having a special test specifically for using this filtering functionality.

Other that I only have some comments here and there and suggestions for maybe more explicit naming, none of which is blocking.

File: crates/astria-sequencer-relayer/src/config.rs
Comment (Fraser999) [reply crates/astria-sequencer-relayer/src/config.rs:130]: Yes, I generally favour that style too.  But in this case, most of the assertions are one-liners, so I didn't think it was worthwhile.

File: crates/astria-sequencer-relayer/src/config.rs
Comment (Fraser999) [reply crates/astria-sequencer-relayer/src/config.rs:133]: I'm thinking probably not, as there's a much greater chance of the `IncludeRollup::new(&input).unwrap()` panicking than the parsing reporting success but actually returning a different set.  If the `unwrap` panics, we don't see the assertion message.

File: crates/astria-sequencer-relayer/local.env.example
Comment (Fraser999) [reply crates/astria-sequencer-relayer/local.env.example:52]: Done in [c745b3f](https://github.com/astriaorg/astria/pull/1001/commits/c745b3f359ac092260ddbeee83289debf9082373).

File: crates/astria-sequencer-relayer/local.env.example
Comment (Fraser999) [reply crates/astria-sequencer-relayer/local.env.example]: Done in [8946615](https://github.com/astriaorg/astria/pull/1001/commits/89466159194d778fe103782251c4bb1920a73c6a).

**Review by joroshiba — APPROVED** [review]
I believe we will need to bump chart version value again, after my merge but looks good to me otherwise.

**Pull request merged**
